src/liballoc/collections/btree/node.rs

```diff
@@ -570,6 +584,7 @@ impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Mut<'a>, K, V, Type> {
         unsafe { &mut *(self.root as *mut Root<K, V>) }
     }
 
+    /// The caller must ensure that this is a proper node, and not the shared root.
```

Comment:
I shouldn't have added the comment without the rest of #67725.
All I'm sure of at the moment is that the current code never feeds shared roots there.

Reply:
When I make a mutable-all-the-way variant of `get_mut`, and Miri says on `test_basic_small` (so without fancy alignment):
```
642 |                     MaybeUninit::first_ptr_mut(&mut (*self.as_leaf_mut()).keys),
    |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Miri evaluation error: Memory access failed: pointer must be in-
bounds at offset 104, but is outside bounds of allocation 22835 which has size 16
    |
```
So as far as I understand, pointers too "must be dereferencable *for the entire size of their pointee*".

Reply:
If I understand correctly the problem is just (as predicted above) that the shared root node has no keys storage but a `&mut` to it where it would be is created, even if only temporarily, and *that* has to be dereferenceable. If it was all raw pointers, you should just get a one-past-the-end pointer (modulo the alignment issue that the `if` fixes) which is never dereferenced and only flows into a zero-sized slice.

But if the other PR removes the support for the shared root from this method anyway, then this is all moot. I think I'd prefer the comment on `as_leaf_mut` to be made more precise and this method to not gain any comment in this PR.

Reply:
I think I'm starting to understand both your comments now. Though if I were to try to correct support for the shared root in `into_key_slice_mut`, I'd still just copy the else-part of `into_key_slice`, kick out some `const`s and splash around some `mut`s until the compiler is happy.

But yes, it's moot, because the callers that want to mutate trees, grant a tree its own root _before_ poking around in slices; and the callers that merely want mutable access to values (i.e. `get_mut`) don't need  `into_key_slice_mut` to handle shared roots.